### PR TITLE
[WIP] Separate plugin API from implementation

### PR DESCRIPTION
**Pull request by dmarcuse**
Moving plugin API stuff into a separate subproject so we can more easily define what should and shouldn't be exposed to plugins, and prevent accidentally using dependencies

Also improving the docs as I go

### CODE REVIEW

File: plugin-api/src/main/java/net/clgd/ccemux/config/JsonAdapter.java [hunk 1]
```diff
@@ -121,10 +121,9 @@ private void fromJson(Group group, JsonElement element) {
 				if (configEntry.get() instanceof Group) {
 					fromJson((Group) configEntry.get(), jsonEntry.getValue());
 				} else if (configEntry.get() instanceof ConfigProperty<?>) {
-					ConfigProperty property = (ConfigProperty<?>) configEntry.get();
```
Comment (SquidDev) [inline plugin-api/src/main/java/net/clgd/ccemux/config/JsonAdapter.java:124]: Does this need to be migrated as well? Seeing as we're going for a grand-unified-config-file, I'm not sure I see the need for exposing it?

File: plugin-api/src/main/java/net/clgd/ccemux/config/package-info.java [hunk 2]
```diff
@@ -0,0 +1,4 @@
+/**
+ * The dynamic config system used by CCEmuX
+ */
+package net.clgd.ccemux.config;
```
Comment (SquidDev) [inline plugin-api/src/main/java/net/clgd/ccemux/config/package-info.java:4]: Please, fix your editor config :).

File: plugin-api/src/main/java/net/clgd/ccemux/emulation/EmuConfig.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+import com.google.gson.Gson;
+import net.clgd.ccemux.config.Config;
+import net.clgd.ccemux.config.JsonAdapter;
+import net.clgd.ccemux.config.ConfigProperty;
+
+/**
+ * The root config used by CCEmuX
+ *
+ */
+public abstract class EmuConfig extends Config {
```
Comment (SquidDev) [inline plugin-api/src/main/java/net/clgd/ccemux/emulation/EmuConfig.java:15]: There's a bit of me which doesn't like exposing this as a concrete class. Aside from the terminal information, we don't really need to expose any of the fields.

File: plugin-api/src/main/java/net/clgd/ccemux/util/OperatingSystem.java [hunk 4]
```diff
@@ -31,4 +37,16 @@ else if (name.startsWith("Mac"))
 		else
 			return Other;
 	}
+
+	public static boolean isWindows() {
+		return get().equals(Windows);
```
Comment (SquidDev) [inline plugin-api/src/main/java/net/clgd/ccemux/util/OperatingSystem.java:42]: I realise this is old code, but if we're following Java conventions the enums should probably be `WINDOWS`, `MAC_OSX`, etc... Seeing as this is an enum, we can also get away with `== Windows` instead, which is a little easier to read.

**Review by SquidDev — CHANGES_REQUESTED** [review]
Looks good. I presume we'd want to move `Plugin` and `hooks.*` into the `plugin-api` directory as well? Though I guess that requires writing some interfaces for the various computer/renderer classes.

File: plugin-api/src/main/java/net/clgd/ccemux/config/JsonAdapter.java (hunk 1 above)
Comment (dmarcuse) [reply plugin-api/src/main/java/net/clgd/ccemux/config/JsonAdapter.java:124]: Probably not, I had just moved the entire `config` package.

File: plugin-api/src/main/java/net/clgd/ccemux/emulation/EmuConfig.java (hunk 3 above)
Comment (dmarcuse) [reply plugin-api/src/main/java/net/clgd/ccemux/emulation/EmuConfig.java:15]: The alternative would be writing getters, which would make it even more verbose to use - `config.getTermScale().get()`. It's not the end of the world, but it's annoying.

**Comment by dmarcuse** [comment]
Yeah, this is still WIP - I'm planning on abstracting some of the emulation-related stuff and then moving `Plugin` and hooks to `plugin-api`, and then also maybe moving the built-in plugins to their own subproject as well.

File: plugin-api/src/main/java/net/clgd/ccemux/emulation/EmuConfig.java (hunk 3 above)
Comment (SquidDev) [reply plugin-api/src/main/java/net/clgd/ccemux/emulation/EmuConfig.java:15]: I guess you could just have `config.termScale().get()` which is slightly shorter. Dammn Java and it's lack of properties :/. I guess I don't really want to include all the "meta" information such as comments and what not.

**Comment by dmarcuse** [comment]
Going to start over from the latest `revival` commit.

**Pull request closed without merging**